stdlib/stdlib.mli

```diff
@@ -51,10 +51,40 @@ exception Exit
 (** The [Exit] exception is not raised by any library function.  It is
     provided for use in your programs. *)
 
-val try_finally : (unit -> 'a) -> (unit -> unit) -> 'a
-(** [try_finally f g] calls [f ()] and returns its result.  If it raises,
-    the same exception is raised unless [g ()] itself raises;
-    in {b: any} case, [g ()] is called after [f ()] terminates.
+val try_finally :
+  ?always:(unit -> unit) ->
+  ?exceptionally:(unit -> unit) ->
+  (unit -> 'a) -> 'a
+(** [try_finally work ~always ~exceptionally] is designed to run code
+    in [work] that may fail with an exception, and has two kind of
+    cleanup routines:
+    {ul
+    {- [always], that must be run after {b any} execution
+       of the function (typically, freeing system resources), and}
+    {- [exceptionally], that should be run {b only} if [work] or [always]
+       failed with an exception (typically, undoing user-visible state
+       changes that would only make sense if the function completes
+       correctly).}
+    }
+    For example:
```

Comment:
This example will get some people to think: I cannot understand it because I don't know the compiler codebase. In fact they probably could, but still it is not engaging. Could you use a realistic but simpler example, or at least using less unknown third-party functions?

Reply:
I agree, but I had not touched it on purpose. I will add a simpler example (if needed) after we settle on an implementation